snapshots/manager.go

```diff
@@ -0,0 +1,223 @@
+func (m *Manager) LoadChunk(height uint64, format uint32, chunk uint32) ([]byte, error) {
+	reader, err := m.store.LoadChunk(height, format, chunk)
+	if err != nil {
+		return nil, err
+	}
+	if reader == nil {
+		return nil, nil
+	}
+	defer reader.Close()
+
+	return ioutil.ReadAll(reader)
+}
+
+// Take takes a snapshot and returns its metadata.
+func (m *Manager) Take(height uint64) (*Snapshot, error) {
+	if m == nil {
+		return nil, errors.New("no snapshot store configured")
+	}
+	err := m.begin(opSnapshot)
+	if err != nil {
+		return nil, err
+	}
+	defer m.end()
+
+	latest, err := m.store.GetLatest()
+	if err != nil {
+		return nil, fmt.Errorf("failed to examine latest snapshot: %w", err)
+	}
+	if latest != nil && latest.Height >= height {
+		return nil, fmt.Errorf("a more recent snapshot already exists at height %v", latest.Height)
+	}
+
+	chunks, err := m.target.Snapshot(height, CurrentFormat)
+	if err != nil {
+		return nil, err
+	}
+	return m.store.Save(height, CurrentFormat, chunks)
+}
+
+// Prune prunes snapshots, if no other operations are in progress.
+func (m *Manager) Prune(retain uint32) (uint64, error) {
+	err := m.begin(opPrune)
+	if err != nil {
+		return 0, err
+	}
+	defer m.end()
+	return m.store.Prune(retain)
+}
+
+// Restore begins an async snapshot restoration, mirroring ABCI OfferSnapshot. Chunks must be fed
+// via RestoreChunk() until the restore is complete or a chunk fails.
+func (m *Manager) Restore(snapshot Snapshot) error {
+	if snapshot.Chunks == 0 {
+		return fmt.Errorf("%w: no chunks", ErrInvalidMetadata)
+	}
+	if uint32(len(snapshot.Metadata.ChunkHashes)) != snapshot.Chunks {
+		return fmt.Errorf("%w: snapshot has %v chunk hashes, but %v chunks",
+			ErrInvalidMetadata,
+			uint32(len(snapshot.Metadata.ChunkHashes)),
+			snapshot.Chunks)
+	}
+	m.mtx.Lock()
+	defer m.mtx.Unlock()
+	err := m.beginLocked(opRestore)
+	if err != nil {
+		return err
+	}
+
+	// Start an asynchronous snapshot restoration, passing chunks and completion status via channels.
+	chChunks := make(chan io.ReadCloser, chunkBufferSize)
+	chDone := make(chan error, 1)
+	go func() {
+		chDone <- m.target.Restore(snapshot.Height, snapshot.Format, chChunks)
+		close(chDone)
+	}()
+
+	// Check for any initial errors from the restore, before any chunks are fed.
+	select {
+	case err := <-chDone:
+		if err == nil {
+			err = errors.New("restore ended unexpectedly")
+		}
+		m.endLocked()
+		return err
+	case <-time.After(20 * time.Millisecond):
+	}
+
+	m.chRestore = chChunks
+	m.chRestoreDone = chDone
+	m.restorePending = snapshot.Metadata.ChunkHashes
+	return nil
+}
+
+// RestoreChunk adds a chunk to an active snapshot restoration, mirroring ABCI ApplySnapshotChunk.
+// Chunks must be given until the restore is complete, returning true, or a chunk errors.
+func (m *Manager) RestoreChunk(chunk []byte) (bool, error) {
+	m.mtx.Lock()
+	defer m.mtx.Unlock()
+	if m.operation != opRestore {
+		return false, fmt.Errorf("no restore operation in progress")
+	}
```

Comment:
how do we know that we are still in the same _restore_ process?

Here is the use-case:

1. Restore started, let's call this process `A` (with process I'm referring to a whole flow, network process, not a goroutine)
1. We received few chunks.
1. in `store.Restore` something fails and we close the `chDone`
1. we start second Restore process (`B`) - this will initialize a new object to `m.chRestoreDone`
1. some node from process `A` will keep sending chunks, then

    1. the `RestoreChunk` function will be called with data from process `A`, but having everything (including `m.chRestoreDone` and `m.chRestore` initialized for process `B`
    2. `store.Restore` will receive data from process `A`
    3. it will probably fail because data doesn't match and we are in a kind of a dead loop.

Reply:
We don't, it's up to the caller to keep track of this. We could return a `Restorer` for the restore operation that the caller has to pass chunks to, but I'm not sure if it's necessary.

Reply:
Also, in your example above, if process A fails, why would it keep sending chunks? It should just exit on error. In this case, the process is Tendermint, so it would be a bug in Tendermint.

Reply:
Process A is in the new node who is performing a restore. 
A node, which offered a snapshot in process A can still send the snapshots, even though A failed. And that chunks will land in B, correct?

Reply:
Tendermint controls the restore process. It will take care of fetching chunks, ordering them, queueing them, and feeding them to the application. If something goes wrong, Tendermint will attempt to start a new restore process. The SDK currently does not support this, since I it's not able to reset its IAVL stores to a fresh state before starting a new restore. This is something we can implement later.

Reply:
I've created this issue: https://github.com/cosmos/cosmos-sdk/issues/7192
Let's add a comment here to mark this error as an `internal` error.